ait/core/server/client.py

```diff
@@ -131,20 +131,74 @@ def __init__(
         zmq_proxy_xpub_url=ait.SERVER_DEFAULT_XPUB_URL,
         **kwargs,
     ):
-        super(PortOutputClient, self).__init__(
+        super(UDPOutputClient, self).__init__(
             zmq_context, zmq_proxy_xsub_url, zmq_proxy_xpub_url
         )
-        self.out_port = kwargs["output"]
+        if "output" in kwargs:
+            output = kwargs["output"]
+            if type(output) is int:
+                self.addr_spec = ("localhost", output)
+            elif utils.is_valid_address_spec(output):
+                protocol, hostname, port = output.split(":")
+                if protocol.lower() != "udp":
+                    raise (
+                        ValueError(f"UDPOutputClient: Invalid Specification {output}")
+                    )
+                self.addr_spec = (hostname, int(port))
+            else:
+                raise (ValueError(f"UDPOutputClient: Invalid Specification {output}"))
+        else:
+            raise (ValueError("UDPOutputClient: Invalid Specification"))
+
         self.context = zmq_context
         # override pub to be udp socket
         self.pub = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
 
     def publish(self, msg):
-        self.pub.sendto(msg, ("localhost", int(self.out_port)))
+        self.pub.sendto(msg, self.addr_spec)
         log.debug("Published message from {}".format(self))
 
 
-class PortInputClient(ZMQClient, gs.DatagramServer):
+class TCPOutputClient(ZMQInputClient):
+    """
+    This is the parent class for all outbound streams which publish
+    to a TCP port. It opens a TCP connection to publish to and publishes
+    outgoing message data to this port.
+    """
+
+    def __init__(
+        self,
+        zmq_context,
+        zmq_proxy_xsub_url=ait.SERVER_DEFAULT_XSUB_URL,
+        zmq_proxy_xpub_url=ait.SERVER_DEFAULT_XPUB_URL,
+        **kwargs,
+    ):
+        super(TCPOutputClient, self).__init__(
+            zmq_context, zmq_proxy_xsub_url, zmq_proxy_xpub_url
+        )
+        if "output" in kwargs:
+            output = kwargs["output"]
+            if utils.is_valid_address_spec(output):
+                protocol, hostname, port = output.split(":")
+                if protocol.lower() != "tcp":
+                    raise (
+                        ValueError(f"TCPOutputClient: Invalid Specification {output}")
+                    )
+                self.addr_spec = (hostname, int(port))
+            else:
+                raise (ValueError(f"TCPOutputClient: Invalid Specification {output}"))
+        else:
+            raise (ValueError("TCPOutputClient: Invalid Specification"))
+
+        self.context = zmq_context
+        self.pub = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+    def publish(self, msg):
+        self.pub.connect(self.addr_spec)
```

Comment:
Just checking if the socket.connect() should be called per-message? Or would it be better to call it once in the constructor?

Also, should re-connection attempt be added in case of broken connection?